[AAQ-81] Retrieval (FAQ matching) validation (#70)
IDinsight/aaq-core: docs/develop/validation.md

# Running validation

> !!! note "Currenlty, there is validation only for retrieval, i.e. `POST /embeddings-search` endpoint."

Reviewer: ```suggestion
!!! note "Currently, there is validation only for retrieval, i.e. `POST /embeddings-search` endpoint."
```

> To evaluate the performance your model (along with your own configurations and

Reviewer: ```suggestion
To evaluate the performance of your model (along with your own configurations and
```

guardrails), run the validation test(s) in `core_backend/validation`.

> The validation is written in pytest so that we can call the endpoint.

Reviewer: I think you can remove this. Don't need to explain design choices to users.

> 
> The validation is written in pytest so that we can call the endpoint.

Author: ```suggestion
```

## Retrieval (`/embeddings-search`) validation

> We evaluate the "performance" of retrieval by computing "Top K Accuracy", which we
> define as proportion of times the best matching answer was present in top K retrieved content.

Reviewer: ```suggestion
We evaluate the "performance" of retrieval by computing "Top K Accuracy", which is
defined as proportion of times the best matching answer was present in top K retrieved contents.
```

### Preparing the data
The test assumes the validation data contains a single label representing the best
matching content, rather than a ranked list of all relevant content.

An example validation data will look like
|query|label|
|--|--|
|"How?"|0|
|"When?"|1|
|"What year was it?"|1|
|"May I?"|2|

An example content data will look like
|content_text|label|
|--|--|
|"Here's how."|0|
|"It was 2024."|1|
|"Yes"|2|

### How to run

1. Create a new python environment:
    ```shell
    conda create -n "aaq-validate" python=3.10
    ```

>     You van also copy the existing `aaq-core` environment.

Reviewer: ```suggestion
    You can also copy the existing `aaq-core` environment.
```

2. Install requirements. This assumes you are in project root `aaq-core`.
    ```shell
    conda activate aaq-validate
    pip install -r core_backend/requirements.txt
    pip install -r core_backend/validation/requirements.txt
    ```
3. Set environment variables.
   The following is required:
    ```
    conda env config vars set OPENAI_API_KEY=<Your OPENAI API key>
    ```
    For optional ones, check out the defaults in `core_backend/app/configs/app_config.py`

>     and modify as per your own requirements:

Reviewer: ```suggestion
    and modify as per your own requirements. For example:
```

    ```
    conda env config vars set EMBEDDING_MODEL=<...>
    ```

> 5. Check the environment variables defined in `core_backend/validation/validation.env`.

Reviewer: Do we really need to check this or defaults just work? Also, what sets up the qdrant container? Is it assumed that there is one already running?

Can I suggest that we just put all of these in a Makefile target like we do for `make tests`. Though i know passing argument is a real pain.

Author: Yeah parsing command line arguments for make target doesn't seem very straightforward.. 

I will not make this because I don't see too much value in making the make target. It's a one-time job to export the required variables to the conda env, and if users are using external API they need to remember to export the token, too (I have clarified this in the docs now!).

   These should all be exported. You can do it one-by-one via `conda env config vars
   set` command or by running
    ```shell
    set -a && source core_backend/validation/validation.env && set +a
    ```
3. In project root `aaq-core` run the following command. (Perform any necessary
   authentication steps you need to do, e.g. for AWS login).
    ```
    python -m pytest core_backend/validation/validate_retrieval.py \
        --validation_data_path <path> \
        --content_data_path <path> \
        --validation_data_question_col <name> \
        --validation_data_label_col <name> \
        --content_data_label_col <name> \
        --content_data_text_col <name> \
        --notification_topic <topic ARN, if using AWS SNS> \
        --aws_profile <aws SSO profile name, if required> \
        -n auto -s
    ```

>     `-n auto` allows multiprocessing to speed up the test, and `-s` ensures logging by

Reviewer: /nit could do this as an annotation. Totally optional.

Author: It is optional but recommended (otherwise too slow) so I will leave it!

Author: -s is also very much recommended to see progress
